_posts/2019-12-05-testowanie-komponentow-z-inputami-i-outputami.md

```diff
@@ -0,0 +1,198 @@
+      declarations: [CalculatorSavedListComponent]
+    }).compileComponents();
+
+    fixture = TestBed.createComponent(CalculatorSavedListComponent);
+    component = fixture.componentInstance;
+    resultText = fixture.debugElement.query(By.css('.result'));
+  }));
+
+  it('should show correct text when list is empty', () => {
+    component.savedValues = [];
+    fixture.detectChanges();
+    expect(resultText.childNodes[0].nativeNode.wholeText).toEqual('Brak zapisanych wartości');
+  });
+
+  it('should show elements of list with correct text', () => {
+    component.savedValues = [12, 3];
+    fixture.detectChanges();
+    expect(resultText.childNodes[0].nativeNode.wholeText).toEqual('Zapisane wartości to: 12,3');
+  });
+});
+```
+
+Zaczęliśmy tutaj, podobnie jak w poprzednim artykule, od stworzenia `TestBed` i dodatkowo wyciągniemy sobie za pomocą zapytania element html prezentujący tekst umieszczony z poziomu komponentu. Możemy do niego odwołać się na wiele sposób, na przykład za pomocą klasy czy id.
+
+Mamy tutaj dwa testy, których struktura jest dokładnie taka sama. Przekazujemy wartości w inpucie i sprawdzamy, czy element umieszczony w html został poprawnie zmieniony. Jeżeli inputów w komponencie mamy więcej, możemy je przetestować w dokładnie ten sam sposób.
+
+## Pisanie testów wprawdzających działanie @Output
+
+Następnie napiszmy test do komponentu pozwalającego wpisać liczbę, którą przekazuje do komponentu nadrzędnego.
+
+```typescript
+@Component({
+  selector: 'calculator-input-field',
+  template: `
+      <input (keyup)="onKey($event)">
+  `,
+})
+export class CalculatorInputFieldComponent {
+  @Output() fieldValue = new EventEmitter<number>();
+
+  onKey(event) {
+    this.fieldValue.emit(event.target.value);
+  }
+}
+```
+
+Komponent równie prosty jak poprzedni z takim wyjątkiem, że mamy tutaj do czynienia z outputem.
+
+```typescript
+describe('CalculatorInputFieldComponent', () => {
+  let component: CalculatorInputFieldComponent;
+  let fixture: ComponentFixture<CalculatorInputFieldComponent>;
+  let input: DebugElement;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [CalculatorInputFieldComponent]
+    }).compileComponents();
+
+    fixture = TestBed.createComponent(CalculatorInputFieldComponent);
+    component = fixture.componentInstance;
+    input = fixture.debugElement.query(By.css('input'));
+  }));
+
+  it('should emit correct value on input change', () => {
+    component.fieldValue.subscribe((value) => expect(value).toBe('7'));
+    input.nativeElement.value = 7;
+    document.querySelector('input').dispatchEvent(new KeyboardEvent('keyup'));
+  });
+});
+```
+
+Czyli standardowo, zaczynamy od przygotowania naszego modułu testowego i uchwycenia elementu, który chcemy przetestować. Warto tutaj zaznaczyć, że istnieją inne sposoby na zainicjowanie komponentu niż `TestBed`, wszystko zależy od  tego co, i w jaki sposób, będziemy testować.
```

Comment:
Warto tutaj zaznaczyć, że istnieją inne sposoby na zainicjowanie komponentu niż `TestBed`, wszystko zależy od  tego co, i w jaki sposób, będziemy testować.

skoro tak piszemy, to warto napisać coś więcej, chociaż jakiś link obrazujący co masz na myśli

Reply:
zastanawiam się, można dać link do jakiegoś artykułu, np https://vsavkin.com/three-ways-to-test-angular-2-components-dcea8e90bd8d , ale nie wiem jaką wartość to będzie miało.

Reply:
> od  tego co, i w jaki sposób

przed tym 'i' nie powinno być przecinka

Reply:
przecinek się znalazł, bo potraktowałem "i w jaki sposób" jako wtrącenie. Możemy olać wszystkie przecinki w tym zdaniu i też będzie sens, ale wydaje mi się, że tak jest też poprawnie